sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/Write.java

```diff
@@ -0,0 +1,255 @@
+  abstract static class Builder<T> {
+    abstract Builder<T> setDataSourceConfiguration(DataSourceConfiguration dataSourceConfiguration);
+
+    abstract Builder<T> setTable(ValueProvider<String> table);
+
+    abstract Builder<T> setBatchSize(ValueProvider<Integer> batchSize);
+
+    abstract Builder<T> setUserDataMapper(UserDataMapper<T> userDataMapper);
+
+    abstract Write<T> build();
+  }
+
+  public Write<T> withDataSourceConfiguration(DataSourceConfiguration config) {
+    checkNotNull(config, "dataSourceConfiguration can not be null");
+    return toBuilder().setDataSourceConfiguration(config).build();
+  }
+
+  public Write<T> withTable(String table) {
+    checkNotNull(table, "table can not be null");
+    return withTable(ValueProvider.StaticValueProvider.of(table));
+  }
+
+  public Write<T> withTable(ValueProvider<String> table) {
+    checkNotNull(table, "table can not be null");
+    return toBuilder().setTable(table).build();
+  }
+
+  public Write<T> withUserDataMapper(UserDataMapper<T> userDataMapper) {
+    checkNotNull(userDataMapper, "userDataMapper can not be null");
+    return toBuilder().setUserDataMapper(userDataMapper).build();
+  }
+
+  /**
+   * Provide a maximum number of rows that is written by one SQL statement. Default is 100000.
+   *
+   * @param batchSize maximum number of rows that is written by one SQL statement
+   */
+  public Write<T> withBatchSize(Integer batchSize) {
+    checkNotNull(batchSize, "batchSize can not be null");
+    return withBatchSize(ValueProvider.StaticValueProvider.of(batchSize));
+  }
+
+  /** Same as {@link #withBatchSize(Integer)} but accepting a ValueProvider. */
+  public Write<T> withBatchSize(ValueProvider<Integer> batchSize) {
+    checkNotNull(batchSize, "batchSize can not be null");
+    return toBuilder().setBatchSize(batchSize).build();
+  }
+
+  @Override
+  public PDone expand(PCollection<T> input) {
+    DataSourceConfiguration dataSourceConfiguration =
+        Util.getRequiredArgument(
+            getDataSourceConfiguration(), "withDataSourceConfiguration() is required");
+    String table = Util.getRequiredArgument(getTable(), "withTable() is required");
+    UserDataMapper<T> userDataMapper =
+        Util.getRequiredArgument(getUserDataMapper(), "withUserDataMapper() is required");
+    int batchSize = Util.getArgumentWithDefault(getBatchSize(), DEFAULT_BATCH_SIZE);
+    checkArgument(batchSize > 0, "batchSize should be greater then 0");
+
+    input
+        .apply(
+            ParDo.of(
+                new DoFn<T, List<String>>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext context) throws Exception {
+                    context.output(userDataMapper.mapRow(context.element()));
+                  }
+                }))
+        .setCoder(ListCoder.of(StringUtf8Coder.of()))
+        .apply(ParDo.of(new BatchFn<>(batchSize)))
+        .apply(ParDo.of(new WriteFn<Void>(dataSourceConfiguration, table)))
```

Comment:
returning void or pdone is considered an antipattern. Better would be to return the status of the writes. This enables users to use the output of this transform to drive future behavior

Reply:
That makes sense. 
I will return a PCollection of integers, where each integer will represent a number of rows written by each LOAD DATA query. It will allow you to check the total number of written rows and see how the task is parallelized.

Reply:
sounds good. Its also frequently useful to return any errors that may have been thrown by the query, especially if those aren't transient